ykcompile/src/lib.rs

```diff
@@ -455,18 +528,9 @@ impl<TT> TraceCompiler<TT> {
                     ; mov [Rq(ro1.reg) + ro1.offs], rax
                 );
             }
-            (Location::Register(reg1), Location::Deref(boxed)) => match *boxed {
-                // For now, referencing a dereference is the same as copying the original pointer.
-                // FIXME: custom Deref implementations.
-                Location::Register(reg2) => {
-                    dynasm!(self.asm
-                        ; mov Rq(reg1), Rq(reg2)
-                    );
-                }
-                _ => todo!(),
-            },
             (_, _) => todo!(),
         };
+        self.free_temporary(rloc);
```

Comment:
I wonder if the localities of these can be improved? It's not immediately obvious to a new reader where the corresponding temp is allocated.

Did you experiment with something that has a `Drop` implementation to automatically free?

Reply:
I have not. I didn't want to overcomplicate this PR. Will we have access to the `register_content_map` if we add a `Drop` implementation to `Location`?

Reply:
`Drop` with side-effects would be very confusing I think.

Reply:
While it's not great that it isn't clear where temporaries are allocated, I agree with @bjorn3 that side-effects on `Drop` would be worse.

Reply:
So `free_temproary` only sometimes frees a temp, right? (i've lost the definition).

Reply:
Yes, it seems passing a location which is not a temp reg is also allowed.

Should this be the case (should it panic?)? If so, should we rename the function to be more indicative of the behaviour? `maybe_free_temp()`?

Reply:
Yes. You give it a `Local` and it's been assigned to a temporary that temporary is freed. Otherwise it does nothing. It's not great, so I'll happily take any suggestions on improving this.

Reply:
We can't `panic` since we will pass it a non temporary a lot. `maybe_free_temp` sounds like it only sometimes frees temporaries. How about something like`free_if_temp`?

Reply:
I think the semantics are fine, just the name could be clearer.

Reply:
`free_if_temp` or `maybe_free_temp` work for me.

Another option might be to implement a `free_temps()` at the statement level and call it after the codegen of each statement. It means we'd only need to call it in one place, but we'd have less control over freeing and would probably leave temps around longer than necessary.

Reply:
We don't have access to `rloc` at the statement level, so we'd have to iterate over all allocated locals after each statement to free the temps.